rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/CompactView.java

```diff
@@ -166,269 +172,107 @@ private void changeSelection(String name) {
+        /* first call : create the components */
+        if (this.normalFlow == null) {
+            initializePanel();
         }
-        return false;
-    }
+        processNodeSources(normalFlow, nodeSources);
 
-    private boolean usedBy(Node n, String username) {
-        return username != null ? username.equals(n.getNodeOwner()) : false;
-    }
+        processNodes(normalFlow, nodes);
 
-    /*
-     * create a copy of the original map containing only resources
-     * used by the current user
-     */
-    private Map<String, NodeSource> filterMyNodes(Map<String, NodeSource> origNodes) {
+        processNodeSources(myNodesFlow, nodeSources);
 
         String username = LoginModel.getInstance().getLogin();
 
-        Map<String, NodeSource> myNodesNsMap = new HashMap<String, NodeSource>();
-        for (String nsid : origNodes.keySet()) {
-            NodeSource origNs = origNodes.get(nsid);
-            NodeSource myNodesNs = null;
-            if (!usedBy(origNs, username)) {
-                continue;
-            } else {
-                myNodesNs = new NodeSource(origNs);
-                myNodesNsMap.put(nsid, myNodesNs);
-            }
+        processNodes(myNodesFlow, nodes.stream().filter(node -> usedBy(node, username)).collect(Collectors.toList()));
 
-            // at this point, at least one node of this ns is being used
-
-            Iterator<String> hostsIt = myNodesNs.getHosts().keySet().iterator();
-            while (hostsIt.hasNext()) {
-                String hostid = hostsIt.next();
-                Host host = myNodesNs.getHosts().get(hostid);
-                if (!usedBy(host, username)) {
-                    hostsIt.remove();
-                    continue;
-                }
+        // if myNodesFlow already have some nodes, but they do not belong to us anymore, then delete them
+        nodes.stream()
+             .filter(node -> !usedBy(node, username))
+             .filter(node -> myNodesFlow.isNodeDrawn(node))
+             .forEach(node -> myNodesFlow.remove(node));
 
-                Iterator<String> nodesIt = host.getNodes().keySet().iterator();
-                while (nodesIt.hasNext()) {
-                    String nodeid = nodesIt.next();
-                    Node node = host.getNodes().get(nodeid);
-                    if (!usedBy(node, username)) {
-                        nodesIt.remove();
-                        continue;
-                    } else {
-                        // node used by the logged user found!
-                    }
-                }
-            }
-        }
-        return myNodesNsMap;
     }
 
-    @Override
-    public void nodesUpdated(Map<String, NodeSource> nodes) {
-        /* show only nodes used by the logged user */
-        if (onlyMyNodes) {
-            nodes = filterMyNodes(nodes);
-        }
+    private boolean usedBy(Node n, String username) {
+        return username != null && username.equals(n.getNodeOwner());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Iterate over the "entrySet" instead of the "keySet". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2864)